File: Java/025_ReverseNodesInKGroup.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 * 先找到下一组的节点，把本组反转后再递归处理后面的节点
 */
public class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode cur = head;
        int number = 0;
        
        while (cur != null && number != k) {
            cur = cur.next;
            number ++;
        }
        
        if (number == k) {
            //找到下一组的头节点
            cur = reverseKGroup(cur, k);
            //将当前组逆向处理，最后记得将最后一个节点处理一下
            while (--number >= 0) {
                ListNode tmp = head.next;
                head.next = cur;
                cur = head;
                head = tmp;
            }
            head = cur;
        }
        
        return head;
    }
}0

File: Python/025_ReverseNodesInKGroup.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        cur = head
        number = 0
        
        while cur and number != k:
            cur = cur.next
            number += 1
        
        if number == k:
            cur = self.reverseKGroup(cur, k)
            
            while number >= 1:
                tmp = head.next
                head.next = cur
                cur = head
                head = tmp
                number -= 1
            
            head = cur
            
        return head;